http_job_queue.go

```diff
@@ -51,163 +53,178 @@ type jobBoardErrorResponse struct {
+	cb *CancellationBroadcaster) (*HTTPJobQueue, error) {
+
 	return &HTTPJobQueue{
-		processorPool:     pool,
-		jobBoardURL:       jobBoardURL,
-		site:              site,
-		providerName:      providerName,
-		queue:             queue,
-		workerID:          workerID,
-		buildJobChanMutex: &sync.Mutex{},
+		jobBoardURL:  jobBoardURL,
+		site:         site,
+		providerName: providerName,
+		queue:        queue,
+		pollInterval: time.Second,
+		cb:           cb,
 	}, nil
 }
 
 // Jobs consumes new jobs from job-board
 func (q *HTTPJobQueue) Jobs(ctx gocontext.Context) (outChan <-chan Job, err error) {
-	q.buildJobChanMutex.Lock()
-	defer q.buildJobChanMutex.Unlock()
-	logger := context.LoggerFromContext(ctx).WithField("self", "http_job_queue")
-	if q.buildJobChan != nil {
-		return q.buildJobChan, nil
-	}
-
 	buildJobChan := make(chan Job)
 	outChan = buildJobChan
+	logger := context.LoggerFromContext(ctx).WithFields(logrus.Fields{
+		"self": "http_job_queue",
+		"inst": fmt.Sprintf("%p", q),
+	})
 
 	go func() {
 		for {
-			logger.Debug("fetching job ids")
-			jobIds, err := q.fetchJobs(ctx)
-			if err != nil {
-				logger.WithField("err", err).Warn("continuing after failing to get job ids")
-				time.Sleep(time.Second)
-			} else {
-				for _, id := range jobIds {
-					logger.WithField("job_id", id).Debug("fetching complete job")
-					buildJob, err := q.fetchJob(ctx, id)
-					if err != nil {
-						logger.WithFields(logrus.Fields{
-							"err": err,
-							"id":  id,
-						}).Warn("failed to get complete job")
-					} else {
-						jobSendBegin := time.Now()
-						buildJobChan <- buildJob
-						metrics.TimeSince("travis.worker.job_queue.http.blocking_time", jobSendBegin)
-						logger.WithFields(logrus.Fields{
-							"source": "http",
-							"dur":    time.Since(jobSendBegin),
-						}).Info("sent job to output channel")
-					}
-				}
-			}
-
-			select {
-			case <-time.After(time.Second):
-				logger.Debug("jobs loop again after 1s sleep")
-				continue
-			case <-ctx.Done():
-				logger.WithField("err", ctx.Err()).Warn("returning from jobs loop due to context done")
-				q.buildJobChan = nil
+			logger.Debug("polling for job tick")
+			if !q.pollForJob(ctx, buildJobChan) {
 				return
 			}
+			time.Sleep(q.pollInterval)
 		}
 	}()
 
-	q.buildJobChan = buildJobChan
 	return outChan, nil
 }
 
-func (q *HTTPJobQueue) fetchJobs(ctx gocontext.Context) ([]uint64, error) {
-	logger := context.LoggerFromContext(ctx).WithField("self", "http_job_queue")
-	fetchRequestPayload := &httpFetchJobsRequest{Jobs: []string{}}
-	numWaiting := 0
-	q.processorPool.Each(func(i int, p *Processor) {
-		switch p.CurrentStatus {
-		case "processing":
-			fetchRequestPayload.Jobs = append(fetchRequestPayload.Jobs, strconv.FormatUint(p.LastJobID, 10))
-		case "waiting", "new":
-			numWaiting++
-		}
+func (q *HTTPJobQueue) pollForJob(ctx gocontext.Context, buildJobChan chan Job) bool {
```

Comment:
I think some docstrings for this and all the other new functions would help make things clearer.